File: tools/confevo/genome.py
"""confevo.genome — Genetic Cargo-feature configuration optimizer: the genome.

This module is the *foundation* of confevo, a tiny genetic algorithm that
searches the space of Cargo feature-sets for the `affidavit` crate. Four sibling
modules (population, fitness, mutation/crossover operators, and the driver)
import the contract defined here, so the names, signatures, and field layout
below are load-bearing — change them and the siblings break.

What is a "genome"?
-------------------
A **genome** is a single *candidate configuration*: a subset of the crate's
Cargo features (see `ALL_FEATURES`). Concretely it is a `frozenset[str]` wrapped
in a frozen dataclass, which makes it immutable, hashable, and usable as a key in
the `seen`/memoization tables the search keeps. Building a genome corresponds to
choosing "which `--features a,b,c` flags would I pass to `cargo`?" — though this
module never invokes cargo or touches the filesystem; it is pure and
deterministic.

Feature implications
--------------------
Cargo features can enable *other* features. `FEATURE_IMPLICATIONS` records, for
each affidavit feature, the set of affidavit-internal features it turns on. For
example `predictive` implies `conformance`, which implies `discovery`, which
implies `core`; and `metrics` implies `otel`. These chains matter because two
genomes that *look* different on the surface can describe the **same** effective
build once implications are applied. `Genome.canonical()` computes the
transitive closure (a fixpoint over `FEATURE_IMPLICATIONS`) so equivalent
configurations collapse to one representative. Callers that want to deduplicate
or compare configurations for *semantic* equality should compare canonical forms
(or their `.key()`), not raw genomes.

Determinism
-----------
Randomness only enters through an explicit `random.Random` instance passed by the
caller (e.g. `Genome.random(rng)`), so seeding the RNG makes every run
reproducible. There is no I/O, no subprocess, and no wall-clock dependence.
"""

from __future__ import annotations

import random
from dataclasses import dataclass

# ---------------------------------------------------------------------------
# Feature universe
# ---------------------------------------------------------------------------

#: Every Cargo feature confevo is allowed to toggle. A `Genome` is a subset of
#: this list; any feature outside it is rejected by validation.
ALL_FEATURES: list[str] = [
    "core", "inspection", "discovery", "conformance", "predictive", "lsp", "file-watch",
    "mutation", "fixture-db", "otel", "metrics", "json-output", "shell", "ui", "benchmarking",
    "profiling", "gpu", "remediation", "pqc", "quality-monitor", "webhook",
]

#: affidavit-internal feature implications: enabling a key feature also enables
#: each feature listed in its value (which may, transitively, enable more). This
#: drives `Genome.canonical()`. Every feature in `ALL_FEATURES` appears here so
#: the table is total; features that imply nothing map to an empty list.
FEATURE_IMPLICATIONS: dict[str, list[str]] = {
    "core": [], "discovery": ["core"], "conformance": ["discovery"],
    "predictive": ["conformance"], "metrics": ["otel"], "profiling": ["benchmarking"],
    "file-watch": ["shell"], "quality-monitor": ["shell"], "webhook": ["shell"],
    # all other features imply nothing internal:
    "inspection": [], "lsp": [], "mutation": [], "fixture-db": [], "otel": [],
    "json-output": [], "shell": [], "ui": [], "benchmarking": [], "gpu": [],
    "remediation": [], "pqc": [],
}

#: Set form of ALL_FEATURES, precomputed for O(1) membership checks.
_ALL_FEATURES_SET: frozenset[str] = frozenset(ALL_FEATURES)


# ---------------------------------------------------------------------------
# Genome
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Genome:
    """A candidate Cargo feature-set.

    The genome is an immutable, hashable wrapper around a `frozenset[str]` of
    feature names. Because the dataclass is frozen and holds a frozenset, value
    equality and hashing come for free: two genomes are `==` iff they hold the
    same raw features.

    Note on equality semantics: raw `==` compares the *literal* feature sets.
    Two genomes that differ only by *implied* features are therefore NOT equal
    as written — but they ARE equal after calling `.canonical()` on both, which
    expands each to its transitive closure. Callers that care about semantic
    (effective-build) equality should compare `a.canonical() == b.canonical()`
    or `a.key() == b.key()`.
    """

    features: frozenset[str]

    def __post_init__(self) -> None:
        # Accept any iterable of feature names but normalize to a frozenset so
        # equality/hashing are well-defined regardless of how the caller built
        # it. Then validate against the known feature universe.
        if not isinstance(self.features, frozenset):
            object.__setattr__(self, "features", frozenset(self.features))
        unknown = self.features - _ALL_FEATURES_SET
        if unknown:
            raise ValueError(
                "unknown feature(s) not in ALL_FEATURES: "
                + ", ".join(sorted(unknown))
            )

    # -- construction -------------------------------------------------------

    @classmethod
    def random(cls, rng: random.Random, p: float = 0.5) -> "Genome":
        """Build a random genome, including each feature independently w.p. `p`.

        `rng` is an explicit `random.Random` so results are reproducible when the
        RNG is seeded. Iterates `ALL_FEATURES` in order so the sequence of draws
        is itself deterministic for a given seed.
        """
        chosen = frozenset(f for f in ALL_FEATURES if rng.random() < p)
        return cls(chosen)

    # -- views --------------------------------------------------------------

    def to_features(self) -> list[str]:
        """Return the genome's features as a sorted list."""
        return sorted(self.features)

    def to_cargo_features_arg(self) -> str:
        """Render as a comma-joined sorted string for `cargo --features`.

        Returns the empty string for the empty genome (i.e. `--no-default-features`
        with nothing added).
        """
        return ",".join(sorted(self.features))

    # -- canonicalization ---------------------------------------------------

    def canonical(self) -> "Genome":
        """Return an equivalent genome with all implied features added.

        Computes the transitive closure over `FEATURE_IMPLICATIONS` via a simple
        fixpoint: keep folding in implications until the set stops growing. The
        result is the genome's *effective* feature-set, so genomes that describe
        the same build canonicalize to equal (and hash-equal) values.
        """
        closure: set[str] = set(self.features)
        while True:
            additions: set[str] = set()
            for feat in closure:
                additions.update(FEATURE_IMPLICATIONS.get(feat, ()))
            if additions <= closure:
                break  # fixpoint reached: nothing new to add
            closure |= additions
        return Genome(frozenset(closure))

    def key(self) -> str:
        """Stable identity string derived from the CANONICAL feature-set.

        Useful as a dict key / dedup token for semantically equal genomes:
        comma-joined sorted canonical features, or the literal "<empty>" when the
        canonical form has no features.
        """
        canon = self.canonical()
        if not canon.features:
            return "<empty>"
        return ",".join(sorted(canon.features))


# ---------------------------------------------------------------------------
# Smoke test
# ---------------------------------------------------------------------------


if __name__ == "__main__":
    rng = random.Random(0)
    genomes = [Genome.random(rng) for _ in range(2)]

    # Also exercise an implication chain explicitly so canonicalization is
    # visible in the output: `predictive` should pull in conformance/discovery/core.
    genomes.append(Genome(frozenset({"predictive", "metrics", "profiling"})))

    for i, g in enumerate(genomes):
        print(f"genome[{i}]")
        print(f"  to_features()            = {g.to_features()}")
        print(f"  to_cargo_features_arg()  = {g.to_cargo_features_arg()!r}")
        print(f"  canonical().to_features()= {g.canonical().to_features()}")
        print(f"  key()                    = {g.key()!r}")

    # Demonstrate the equality contract: differs only by implied features.
    bare = Genome(frozenset({"predictive"}))
    expanded = Genome(frozenset({"predictive", "conformance", "discovery", "core"}))
    print()
    print(f"bare == expanded                         : {bare == expanded}")
    print(f"bare.canonical() == expanded.canonical() : {bare.canonical() == expanded.canonical()}")
    print(f"bare.key() == expanded.key()             : {bare.key() == expanded.key()}")

File: tools/confevo-rs/src/genome.rs
//! The [`Genome`]: a single candidate feature configuration.
//!
//! A genome is an immutable, hashable subset of a [`FeatureSpace`]'s universe —
//! "which `--features a,b,c` would I pass to `cargo`?". It is pure: constructing,
//! crossing, and mutating genomes never invokes cargo or touches the filesystem.
//!
//! Two genomes that differ only by *implied* features are not literally equal, but
//! they share a [`Genome::key`] (derived from the canonical closure), which is how
//! the search deduplicates semantically-identical configurations.

use std::collections::BTreeSet;

use crate::rng::Rng;
use crate::space::FeatureSpace;

/// A candidate Cargo feature-set: an immutable subset of a [`FeatureSpace`].
#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct Genome {
    feats: BTreeSet<String>,
}

impl Genome {
    /// Build a genome from any iterable of feature names.
    ///
    /// Names are stored verbatim (deduplicated by the `BTreeSet`); validity
    /// against a space is the caller's concern — the GA only ever produces genomes
    /// drawn from the space, and [`Genome::canonical`]/[`Genome::key`] ignore
    /// out-of-space names.
    pub fn new<I, S>(feats: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        Genome {
            feats: feats.into_iter().map(Into::into).collect(),
        }
    }

    /// The empty genome (`cargo build --no-default-features` with nothing added).
    pub fn empty() -> Self {
        Genome {
            feats: BTreeSet::new(),
        }
    }

    /// Build a random genome, including each feature in `space` independently with
    /// probability `p`.
    ///
    /// Iterates `space.features()` in order so the draw sequence — and therefore
    /// the genome — is reproducible for a seeded [`Rng`].
    pub fn random(rng: &mut Rng, space: &FeatureSpace, p: f64) -> Self {
        let mut feats = BTreeSet::new();
        for f in space.features() {
            if rng.gen_bool(p) {
                feats.insert(f.clone());
            }
        }
        Genome { feats }
    }

    /// The genome's features as a sorted `Vec`.
    pub fn features(&self) -> Vec<String> {
        self.feats.iter().cloned().collect()
    }

    /// Borrow the underlying feature set.
    pub fn feature_set(&self) -> &BTreeSet<String> {
        &self.feats
    }

    /// Number of features enabled in this genome (literal, not canonical).
    pub fn len(&self) -> usize {
        self.feats.len()
    }

    /// `true` if no features are enabled.
    pub fn is_empty(&self) -> bool {
        self.feats.is_empty()
    }

    /// `true` if `feature` is enabled in this genome (literal membership).
    pub fn contains(&self, feature: &str) -> bool {
        self.feats.contains(feature)
    }

    /// Render as a comma-joined sorted string for `cargo --features`.
    ///
    /// Returns the empty string for the empty genome.
    pub fn cargo_features_arg(&self) -> String {
        self.features().join(",")
    }

    /// Return an equivalent genome with all implied features added.
    ///
    /// Delegates to [`FeatureSpace::closure`], so the result is the genome's
    /// *effective* feature-set under `space`'s implication edges.
    pub fn canonical(&self, space: &FeatureSpace) -> Genome {
        Genome {
            feats: space.closure(&self.feats),
        }
    }

    /// Stable identity string derived from the **canonical** feature-set.
    ///
    /// Used as the dedup / memoization token: comma-joined sorted canonical
    /// features, or the literal `"<empty>"` when the canonical form is empty. Two
    /// genomes describing the same effective build share a key.
    pub fn key(&self, space: &FeatureSpace) -> String {
        let canon = self.canonical(space);
        if canon.feats.is_empty() {
            "<empty>".to_string()
        } else {
            canon.features().join(",")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn demo_space() -> FeatureSpace {
        // predictive -> conformance -> discovery -> core ; metrics -> otel
        FeatureSpace::new(
            [
                "core",
                "discovery",
                "conformance",
                "predictive",
                "otel",
                "metrics",
                "ui",
            ],
            [
                ("discovery", vec!["core"]),
                ("conformance", vec!["discovery"]),
                ("predictive", vec!["conformance"]),
                ("metrics", vec!["otel"]),
            ],
        )
        .unwrap()
    }

    #[test]
    fn cargo_arg_is_sorted_and_comma_joined() {
        let g = Genome::new(["ui", "core", "otel"]);
        assert_eq!(g.cargo_features_arg(), "core,otel,ui");
    }

    #[test]
    fn empty_genome_renders_blank_and_keys_as_empty() {
        let space = demo_space();
        let g = Genome::empty();
        assert_eq!(g.cargo_features_arg(), "");
        assert_eq!(g.key(&space), "<empty>");
    }

    #[test]
    fn canonical_pulls_in_implication_chain() {
        let space = demo_space();
        let g = Genome::new(["predictive"]);
        assert_eq!(
            g.canonical(&space).features(),
            vec!["conformance", "core", "discovery", "predictive"]
        );
    }

    #[test]
    fn genomes_differing_only_by_implications_share_a_key() {
        let space = demo_space();
        let bare = Genome::new(["predictive"]);
        let expanded = Genome::new(["predictive", "conformance", "discovery", "core"]);
        assert_ne!(bare, expanded); // literally different
        assert_eq!(bare.key(&space), expanded.key(&space)); // same effective build
    }

    #[test]
    fn random_is_reproducible_for_a_seed() {
        let space = demo_space();
        let mut a = Rng::new(7);
        let mut b = Rng::new(7);
        assert_eq!(
            Genome::random(&mut a, &space, 0.5),
            Genome::random(&mut b, &space, 0.5)
        );
    }

    #[test]
    fn random_p_bounds_are_total_and_empty() {
        let space = demo_space();
        let mut rng = Rng::new(1);
        assert_eq!(Genome::random(&mut rng, &space, 1.0).len(), space.len());
        assert!(Genome::random(&mut rng, &space, 0.0).is_empty());
    }
}
